dhall/src/Dhall/ETags.hs

```diff
@@ -0,0 +1,216 @@
+{-# LANGUAGE CPP #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE LambdaCase #-}
+
+-- | This module contains the implementation of the @dhall tags@ command
+
+module Dhall.ETags
+    ( -- * ETags
+      generate
+    ) where
+
+#if !(MIN_VERSION_base(4,9,0))
+import Control.Applicative (liftA2)
+#endif
+import Control.Exception (handle, SomeException(..))
+#if MIN_VERSION_base(4,10,0)
+import Data.Either (fromRight)
+#endif
+import Data.List (isSuffixOf)
+import Data.Maybe (fromMaybe)
+import Data.Semigroup (Semigroup(..))
+import Dhall.Map (foldMapWithKey)
+import Data.Text (Text)
+import Dhall.Util (Input(..))
+import Dhall.Core (Expr(..), Binding(..))
+import Dhall.Src (Src(srcStart))
+import Dhall.Parser (exprFromText)
+import System.Directory ( doesDirectoryExist
+#if MIN_VERSION_directory(1,3,0)
+                        , pathIsSymbolicLink
+#elif MIN_VERSION_directory(1,2,6)
+                        , isSymbolicLink
+#endif
+                        , getDirectoryContents
+                        )
+import System.FilePath ((</>), takeFileName)
+import Text.Megaparsec (sourceLine, sourceColumn, unPos)
+
+import qualified Data.Map      as M
+import qualified Data.Text     as T
+import qualified Data.Text.IO  as TIO
+
+#if !(MIN_VERSION_base(4,9,0))
+instance Monoid a => Monoid (IO a) where
+    mempty = pure mempty
+    mappend = liftA2 mappend
+#endif
```

Comment:
Sorry, I wasn't aware that the instance isn't available with GHC-7.10. Orphan instances like these are a rather problematic, because anyone who imports this module will suddenly also have this instance in scope.

In general, I think it's also good to use CPP as sparingly as possible.

Reply:
okay, removed them all.

Reply:
Except the ones, I need to check if path is symlink.

Reply:
Cheers! :)